laikaIO/src/main/scala/pink/cozydev/protosearch/analysis/PlaintextRenderer.scala

```diff
@@ -16,40 +16,123 @@
 
 package pink.cozydev.protosearch.analysis
 
-import laika.ast._
+import laika.ast.*
 import laika.api.format.{Formatter, RenderFormat}
+import laika.ast.html.{HTMLBlock, HTMLSpan}
 
 object PlaintextRenderer extends ((Formatter, Element) => String) {
 
-  private case class Content(content: Seq[Element], options: Options = Options.empty)
-      extends Element
-      with ElementContainer[Element] {
-    type Self = Content
-    def withOptions(options: Options): Content = copy(options = options)
-  }
-
   def apply(fmt: Formatter, element: Element): String = {
 
-    def renderElement(e: Element): String = {
-      val (elements, _) = e.productIterator.partition(_.isInstanceOf[Element])
-      e.productPrefix + fmt.indentedChildren(
-        elements.toList.asInstanceOf[Seq[Element]]
-      )
+    def renderElement(e: Element): String = e match {
+
+      /* search engines tend to index alt and title attributes of images */
+      case img: Image => (img.alt.toList ++ img.title.toList).mkString(" ")
```

Comment:
(not a change request, just me thinking outloud)

This line of code made me realize a cheap and cheerful way to identify "fragments" within the "plain text" in order to divide the text up into logical pieces during highlighting (https://github.com/cozydev-pink/protosearch/issues/205). Which is to say, delimiting "fragments" by newlines. Effectively making every line in the plain text format a "fragment" and then, during highlighting we figure out the fragment that best matches the query.
So in some future effort I might change this line to use `.mkString("\n")`, making the caption and the title to different fragments to potentially highlight. Or maybe not, maybe it's better to have them together in one fragment. It's just nice that there is a pretty easy way to control this in this renderer, if we go that route.

Reply:
Yes, I think you're right about needing to reparse.
Because rendering in Laika requires us to output `String`, we're a bit limited in how we can preserve structure without requiring an additional parsing step.
But that's not the end of the world or anything. We *have* to tokenize the text anyways, so ideally we can handle the secondary parsing there.